File: Reto # - 8/Python/Clases.py
class MyClass:
    def __init__(self, param1, param2):
        self.param1 = param1
        self.param2 = param2
    
    def print_params(self):
        print("Param 1:", self.param1)
        print("Param 2:", self.param2)

# Crear instancia de la clase
my_object = MyClass("Value 1", "Value 2")

# Imprimir los parámetros
my_object.print_params()

# Modificar los parámetros
my_object.param1 = "New Value 1"
my_object.param2 = "New Value 2"

# Imprimir los parámetros modificados
my_object.print_params()


class Stack:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return len(self.items) == 0

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if not self.is_empty():
            return self.items.pop()
        else:
            print("La pila está vacía")

    def peek(self):
        if not self.is_empty():
            return self.items[-1]
        else:
            print("La pila está vacía")

    def size(self):
        return len(self.items)

    def print_stack(self):
        print("Contenido de la pila:")
        for item in reversed(self.items):
            print(item)


# Ejemplo de uso de la clase Pila
stack = Stack()
stack.push(1)
stack.push(2)
stack.push(3)

stack.print_stack()

print("Tamaño de la pila:", stack.size())
print("Elemento superior de la pila:", stack.peek())

print("Elemento eliminado de la pila:", stack.pop())
stack.print_stack()

class Queue:
    def __init__(self):
        self.items = []

    def is_empty(self):
        return len(self.items) == 0

    def enqueue(self, item):
        self.items.append(item)

    def dequeue(self):
        if not self.is_empty():
            return self.items.pop(0)
        else:
            print("La cola está vacía")

    def size(self):
        return len(self.items)

    def print_queue(self):
        print("Contenido de la cola:")
        for item in self.items:
            print(item)


# Ejemplo de uso de la clase Cola
queue = Queue()
queue.enqueue(1)
queue.enqueue(2)
queue.enqueue(3)

queue.print_queue()

print("Tamaño de la cola:", queue.size())

print("Elemento eliminado de la cola:", queue.dequeue())
queue.print_queue()

File: Reto # - 8/C++/Clases.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

class MyClass {
private:
    std::string param1;
    std::string param2;

public:
    MyClass(std::string param1, std::string param2) : param1(param1), param2(param2) {}

    void printParams() {
        std::cout << "Param 1: " << param1 << std::endl;
        std::cout << "Param 2: " << param2 << std::endl;
    }
};


class Stack {
private:
    std::vector<int> items;

public:
    bool isEmpty() {
        return items.empty();
    }

    void push(int item) {
        items.push_back(item);
    }

    int pop() {
        if (!isEmpty()) {
            int top = items.back();
            items.pop_back();
            return top;
        } else {
            std::cout << "La pila está vacía" << std::endl;
            return -1; // Retorno de valor por defecto en caso de pila vacía
        }
    }

    int peek() {
        if (!isEmpty()) {
            return items.back();
        } else {
            std::cout << "La pila está vacía" << std::endl;
            return -1; // Retorno de valor por defecto en caso de pila vacía
        }
    }

    int size() {
        return items.size();
    }

    void printStack() {
        std::cout << "Contenido de la pila:" << std::endl;
        for (auto it = items.rbegin(); it != items.rend(); ++it) {
            std::cout << *it << std::endl;
        }
    }
};


class Queue {
private:
    std::queue<int> items;

public:
    bool isEmpty() {
        return items.empty();
    }

    void enqueue(int item) {
        items.push(item);
    }

    int dequeue() {
        if (!isEmpty()) {
            int front = items.front();
            items.pop();
            return front;
        } else {
            std::cout << "La cola está vacía" << std::endl;
            return -1; // Retorno de valor por defecto en caso de cola vacía
        }
    }

    int size() {
        return items.size();
    }

    void printQueue() {
        std::cout << "Contenido de la cola:" << std::endl;
        std::queue<int> temp = items;
        while (!temp.empty()) {
            std::cout << temp.front() << std::endl;
            temp.pop();
        }
    }
};



int main() {
    // Crear instancia de la clase
    MyClass myObject("Value 1", "Value 2");

    // Imprimir los parámetros
    myObject.printParams();

    // Modificar los parámetros
    myObject.param1 = "New Value 1";
    myObject.param2 = "New Value 2";

    // Imprimir los parámetros modificados
    myObject.printParams();


    //Stack
    Stack stack;
    stack.push(1);
    stack.push(2);
    stack.push(3);

    stack.printStack();

    std::cout << "Tamaño de la pila: " << stack.size() << std::endl;
    std::cout << "Elemento superior de la pila: " << stack.peek() << std::endl;

    std::cout << "Elemento eliminado de la pila: " << stack.pop() << std::endl;
    stack.printStack();

    //Queue

    Queue queue;
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);

    queue.printQueue();

    std::cout << "Tamaño de la cola: " << queue.size() << std::endl;

    std::cout << "Elemento eliminado de la cola: " << queue.dequeue() << std::endl;
    queue.printQueue();

    return 0;
}